samples/05.custom-components/i.notification/README.md

```diff
@@ -0,0 +1,346 @@
++   hooks: { useDismissNotification, useSendPostBack, useSetNotification }
++ } = window.WebChat;
+
+  const PrivacyPolicyToast = () => {
++   const sendPostBack = useSendPostBack();
++   const dismissNotification = useDismissNotification();
++   const setNotification = useSetNotification();
++
++   const handleDismissNotification = useCallback(() => dismissNotification('privacypolicy'), [
++     dismissNotification
++   ]);
++
++   const handleAgreeClick = useCallback(() => {
++     sendPostBack({ acceptPrivacyPolicy: true });
++
++     setNotification({
++       data: { accepted: true },
++       id: 'privacypolicy',
++       level: 'success',
++       message: 'Thanks for accepting our [privacy policy](https://privacy.microsoft.com/).'
++     });
++   }, []);
+
+    return (
+      <div aria-label="Privacy policy" role="dialog" className="app__privacyPolicyToast">
+        <i aria-hidden={true} className="ms-Icon ms-Icon--Compare app__privacyPolicyToast__icon" />
+        <span>
+          Do you consent to our{' '}
+          <a href="https://privacy.microsoft.com/" rel="noopener noreferrer" target="_blank">
+            privacy policy
+          </a>
+          ?
+        </span>
+-       <button className="app__privacyPolicyToast__button" type="button">
++       <button className="app__privacyPolicyToast__button" onClick={handleAgreeClick} type="button">
+          Yes
+        </button>{' '}
+-       <button className="app__privacyPolicyToast__button" type="button">
++       <button className="app__privacyPolicyToast__button" onClick={handleDismissNotification} type="button">
+          No
+        </button>
+      </div>
+    );
+  };
+```
+
+Lastly, we will add a toast middleware to intercept the render of toasts. The middleware will override the rendering for toast with ID `privacypolicy` and is not accepted by the end-user yet.
+
+When end-user give consent to the privacy policy, we will use the default renderer to render the toast.
+
+```diff
++ const toastMiddleware = () => next => ({ notification, ...otherArgs }) => {
++   if (notification.id === 'privacypolicy' && !notification.data.accepted) {
++     return <PrivacyPolicyToast />;
++   }
++
++   return next({ notification, ...otherArgs });
++ };
+
+  window.WebChat.renderWebChat(
+    {
+      directLine: window.WebChat.createDirectLine({ token }),
++     toastMiddleware,
+      store
+    },
+    document.getElementById('webchat')
+  );
+```
+
+## Completed code
+
+Here is the finished `index.html`:
+
+```diff
+  const res = await fetch('https://webchat-mockbot.azurewebsites.net/directline/token', { method: 'POST' });
+  const { token } = await res.json();
++ const { useCallback, useMemo } = window.React;
++ const {
++   hooks: { useDismissNotification, useSendPostBack, useSetNotification }
++ } = window.WebChat;
++
++ const store = window.WebChat.createStore({}, () => next => action => {
++   if (action.type === 'DIRECT_LINE/CONNECT_FULFILLED') {
++     store.dispatch({
```

Comment:
Is there a reason why you're not using the `setNotification` function here? Also, you can use the `dispatch` function from the store middleware's parameters instead of `store.dispatch`.

```js
window.WebChat.createStore(({ dispatch }) => next => action => {
    if (action.type === 'DIRECT_LINE/CONNECT_FULFILLED') {
       dispatch({ ... })
    }
}
```

Reply:
If you mean `setNotification` action creator, heh, we didn't expose it from Web Chat core to `webchat.js`.

If you mean `useSetNotification` hook, we need to put it under `<Composer>` in order to work, that would complicate the sample.

Eventually would love to expose action creators.

Will use the `dispatch`. Thanks for the tips!